### PR TITLE
KYLIN-4966 Refresh the existing segment according to the new cuboid list in kylin4

### PR DESCRIPTION
**Pull request by zhangayqian**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Kylin?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have create an issue on [Kylin's jira](https://issues.apache.org/jira/browse/KYLIN), and have described the bug/feature there in detail
- [ ] Commit messages in my PR start with the related jira ID, like "KYLIN-0000 Make Kylin project open-source"
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If this change need a document change, I will prepare another pr against the `document` branch
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at user@kylin or dev@kylin by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zhangayqian** [comment]
## Test Evidence

- Set `kylin.engine.segment-statistics-enabled=true`

- Build two segment
image

- Call the API to modify cuboid list and submit optimizing job
image

image

- Optimize Job
image

image

image


- Checkpoint Job
image

- After all jobs complete，cube segment meta have been updated
image

- Cuboid file in hdfs
image

- Query
image

File: kylin-spark-project/kylin-spark-engine/src/main/java/org/apache/kylin/engine/spark/job/FilterRecommendCuboidJob.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            for (FileStatus cuboid : fs.listStatus(originalCuboidPath)) {
+                String cuboidId = cuboid.getPath().getName();
+                if (cuboidId.equals(String.valueOf(baseCuboid)) || recommendCuboids.contains(Long.valueOf(cuboidId))) {
+                    Path optimizeCuboidPath = new Path(getCuboidRootPath(optimizeSegment) + "/" + cuboidId);
+                    FileUtil.copy(fs, cuboid.getPath(), fs, optimizeCuboidPath, false, true, conf);
+                    logger.info("Copy cuboid {} storage from original segment to optimized segment", cuboidId);
+                }
+            }
+        } catch (IOException e) {
+            logger.error("Failed to filter cuboid", e);
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-engine/src/main/java/org/apache/kylin/engine/spark/job/FilterRecommendCuboidJob.java:90]: Could you please add cleanup step in catch clause? Maybe remove file under `optimizeCuboidPath` ? Or add cleanup task to StorageCleanupJob ?

File: kylin-spark-project/kylin-spark-engine/src/main/java/org/apache/kylin/engine/spark/job/FilterRecommendCuboidJob.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import org.apache.kylin.engine.spark.metadata.cube.PathManager;
+import org.apache.kylin.shaded.com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Set;
+
+public class FilterRecommendCuboidJob extends SparkApplication {
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-engine/src/main/java/org/apache/kylin/engine/spark/job/FilterRecommendCuboidJob.java:42]: The class name could be more meaningful.

Comment (hit-lacus) [reply kylin-spark-project/kylin-spark-engine/src/main/java/org/apache/kylin/engine/spark/job/FilterRecommendCuboidJob.java:42]: Maybe `CopyRemainingCuboidJob` ?

File: core-job/src/main/java/org/apache/kylin/job/constant/ExecutableConstants.java [hunk 3]
```diff
@@ -94,5 +94,6 @@ private ExecutableConstants() {
     public static final String FLINK_SPECIFIC_CONFIG_NAME_MERGE_DICTIONARY = "mergedict";
     //kylin on parquetv2
     public static final String STEP_NAME_DETECT_RESOURCE = "Detect Resource";
+    public static final String STEP_NAME_BUILD_CUBOID_FROM_PARENT_CUBOID = "Build recommend cuboid from parent cuboid";
```
Comment (hit-lacus) [inline core-job/src/main/java/org/apache/kylin/job/constant/ExecutableConstants.java:97]: Use `exist` or `parent` ?

File: kylin-spark-project/kylin-spark-engine/src/main/java/org/apache/kylin/engine/spark/job/FilterRecommendCuboidJob.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        CubeSegment oldSegment = optimizeSegment.getCubeInstance().getOriginalSegmentToOptimize(optimizeSegment);
+        Preconditions.checkNotNull(oldSegment,
+                "cannot find the original segment to be optimized by " + optimizeSegment);
+
+        infos.recordReusedCuboids(Collections.singleton(cube.getCuboidsByMode(CuboidModeEnum.RECOMMEND_EXISTING)));
+
+        baseCuboid = cube.getCuboidScheduler().getBaseCuboidId();
+        recommendCuboids = cube.getCuboidsRecommend();
+
+        Preconditions.checkNotNull(recommendCuboids, "The recommend cuboid map could not be null");
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-engine/src/main/java/org/apache/kylin/engine/spark/job/FilterRecommendCuboidJob.java:76]: `map` or `set`?

File: kylin-spark-project/kylin-spark-engine/src/main/java/org/apache/kylin/engine/spark/job/NSparkBatchOptimizeJobCheckpointBuilder.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        Preconditions.checkNotNull(cube.getFirstSegment(), "Cube " + cube + " is empty!!!");
+    }
+
+    public CheckpointExecutable build() {
+        KylinConfig kylinConfig = cube.getConfig();
+        List<ProjectInstance> projList = ProjectManager.getInstance(kylinConfig).findProjects(cube.getType(),
+                cube.getName());
+        if (projList == null || projList.size() == 0) {
+            throw new RuntimeException("Cannot find the project containing the cube " + cube.getName() + "!!!");
+        } else if (projList.size() >= 2) {
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-engine/src/main/java/org/apache/kylin/engine/spark/job/NSparkBatchOptimizeJobCheckpointBuilder.java:54]: In which case, could `projList.size() >= 2` happen ?

File: kylin-spark-project/kylin-spark-engine/src/main/java/org/apache/kylin/engine/spark/job/NSparkBatchOptimizeJobCheckpointBuilder.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        // Phase 1: Update cube information
+        checkpointJob.addTask(createUpdateCubeInfoAfterCheckpointStep());
+
+        // Phase 2: Cleanup hdfs storage
+        checkpointJob.addTask(createCleanupHdfsStorageStep());
+
+        return checkpointJob;
+    }
+
+    private NSparkUpdateCubeInfoAfterOptimizeStep createUpdateCubeInfoAfterCheckpointStep() {
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-engine/src/main/java/org/apache/kylin/engine/spark/job/NSparkBatchOptimizeJobCheckpointBuilder.java:75]: `AfterCheckpointStep` or `AfterOptimizeJob` ?

File: kylin-spark-project/kylin-spark-engine/src/main/java/org/apache/kylin/engine/spark/job/NSparkOptimizingStep.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    public static class Mockup {
+        public static void main(String[] args) {
+            logger.info(NSparkCubingStep.Mockup.class + ".main() invoked, args: " + Arrays.toString(args));
+        }
+    }
+
+    @Override
+    public boolean needMergeMetadata() {
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-engine/src/main/java/org/apache/kylin/engine/spark/job/NSparkOptimizingStep.java:67]: Why overwrite this method?

**Review by hit-lacus — APPROVED** [review]
LGTM. Let's merge this patch at the moment.

**Pull request merged**